packages/remark-abbr/lib/index.js

```diff
@@ -1,152 +1,247 @@
-import { visit } from 'unist-util-visit'
-
-export default function plugin (options) {
-  const opts = options || {}
-  const expandFirst = opts.expandFirst
-
-  function locator (value, fromIndex) {
-    return value.indexOf('*[', fromIndex)
+import { SKIP, CONTINUE, visit } from 'unist-util-visit'
+import { abbr, abbrTypes } from 'micromark-extension-abbr'
+
+function splitTextByAbbr (textNode, abbrDefinitions, seenAbbreviations, opts) {
+  const expandFirst = opts.expandFirst || false
+  const uniqueAbbreviationMap = new Map()
+  for (const abbreviation of abbrDefinitions) {
+    uniqueAbbreviationMap.set(abbreviation.identifier, abbreviation)
   }
 
-  function inlineTokenizer (eat, value, silent) {
-    const regex = /[*]\[([^\]]*)\]:\s*(.+)\n*/
-    const keep = regex.exec(value)
+  const uniqueAbbreviations = [...uniqueAbbreviationMap.values()]
+
+  const matches = uniqueAbbreviations
+    .map(
+      (abbr) =>
+        /** @type {const} */ ([abbr, textNode.value.indexOf(abbr.identifier)])
+    )
+    .filter(([_abbr, index]) => index >= 0)
+    .map(([abbr, index]) => {
+      const start = index
+      const end = index + abbr.identifier.length - 1
+      return {
+        abbr,
+        start,
+        end,
+        prevChar: textNode.value[start - 1],
+        nextChar: textNode.value[end + 1]
+      }
+    })
+    .filter((match) =>
+      // We don't want to match "HTML" inside strings like "HHHHTMLLLLLL", so check that the
+      // surrounding characters are either undefined (i.e. start of string / end of string)
+      // or non-word characters
+      [match.prevChar, match.nextChar].every(
+        (c) => c === undefined || /^\W$/.test(c)
```

Comment:
I disagree with `\W` because it does not work outside of ASCII characters. For example try the plugin with abbr `*[НПО]: неправительственная организация` (russian for NGO) and input text `НННПООО` (same thing as what you did for `HTML`, I doubled the letters). The trailing `Н` (not latin H) and leading `О` (not latin O) are taken for non-word characters. But they are word characters, in another language. I suggest to use a proper plugin instead, I think we use something similar somewhere, search for `unicode` in project dependencies.